File: skills/docforge/_shared/runtime/validation/validate_metadata.py
#!/usr/bin/env python3
"""Validate the Docforge catalog, schemas, templates, peers, and release metadata."""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

from runtime.common._util import read_json
from runtime.common.provenance_frontmatter import PROVENANCE_FIELDS, SCHEMA_VERSION, parse_frontmatter
from . import generate_indexes
from runtime.catalog import query_catalog

SKILL_ROOT = Path(__file__).resolve().parent.parent.parent
REPO_ROOT = SKILL_ROOT.parent.parent
CATALOG_VERSION = "2.6.1"
MARKDOWN_EXCEPTIONS = {"agents-kernel.md", "claude-md.md", "claude-local-md.md"}
PUBLIC_CONTRACTS = {
    "manage_manifest": ["init", "add", "set", "status", "audit", "--repo", "--tier", "--shape", "--platform", "--framework", "--concern", "--audience", "--type", "--id", "--path", "--status", "--mode", "--verdict", "--report"],
    "detect_profiles": ["--repo", "--json", "--emit-gate-pack", "confirmed", "candidate"],
    "scaffold_docs": ["--repo", "--manifest", "--dry-run", "--document", "--audit"],
    "precheck_graph": ["--repo", "--need", "code", "flow"],
    "check_staleness": ["--manifest", "--document", "--section", "--json", "--sync-provenance"],
    "flow_index": ["harvest", "revise", "render", "organize", "emit", "apply", "--repo", "--gitnexus-export", "--main-limit", "--output", "--organization"],
    "migrate_metadata": ["--repo", "--manifest", "--dry-run", "--report"],
    "query_catalog": ["--tier", "--id", "--ids", "--profile", "--applicable", "--validate", "--category", "--route"],
    "generate_indexes": ["--write", "--check"],
}


def validate() -> list[str]:
    errors: list[str] = []
    metadata = SKILL_ROOT / ".metadata"
    # Catalog shape checks live in query_catalog --validate (single source of truth).
    errors.extend(query_catalog.validate())
    try:
        stale = generate_indexes.stale_targets()
        for path in stale:
            errors.append(f"generated router is stale: {path.relative_to(SKILL_ROOT)} (run generate_indexes --write)")
    except ValueError as exc:
        errors.append(f"generate_indexes --check failed: {exc}")
    try:
        catalog = query_catalog.as_legacy_catalog()
    except ValueError as exc:
        errors.append(str(exc))
        catalog = {"version": None, "documents": [], "profiles": {}, "tiers": []}
    catalog_schema = read_json(metadata / "catalog-schema.json")
    manifest_schema = read_json(metadata / "manifest-schema.json")
    flow_index_schema = read_json(metadata / "flow-index-schema.json")
    provenance_schema_path = metadata / "provenance-schema.json"
    if not provenance_schema_path.is_file():
        errors.append("provenance-schema.json is missing")
    else:
        provenance_schema = read_json(provenance_schema_path)
        if provenance_schema.get("properties", {}).get("schema", {}).get("const") != SCHEMA_VERSION:
            errors.append("provenance schema must require schema 2.0")
    if catalog.get("version") != CATALOG_VERSION:
        errors.append(f"catalog version must be {CATALOG_VERSION}")
    if catalog_schema.get("properties", {}).get("version", {}).get("const") != CATALOG_VERSION:
        errors.append("catalog schema version disagrees with catalog")
    if not (metadata / "catalog" / "index.json").is_file():
        errors.append("split catalog index.json is missing")
    if (metadata / "catalog.json").is_file():
        errors.append("obsolete monolith catalog.json remains; use .metadata/catalog/")
    if manifest_schema.get("properties", {}).get("version", {}).get("const") != "3.1":
        errors.append("manifest schema must require version 3.1")
    if flow_index_schema.get("properties", {}).get("version", {}).get("const") != "1.1":
        errors.append("flow index schema must require version 1.1")
    flow_item = (
        flow_index_schema.get("properties", {})
        .get("flows", {})
        .get("items", {})
        .get("properties", {})
    )
    for field in ("display_name", "family", "doc_role", "composed_into", "doc_path"):
        if field not in flow_item:
            errors.append(f"flow index schema must define flow.{field}")
    if flow_item.get("doc_role", {}).get("enum") != ["standalone", "member", "index_only"]:
        errors.append("flow index schema doc_role must be standalone|member|index_only")
    dimensions = ["shapes", "platforms", "frameworks", "concerns", "audiences"]
    profiles = catalog.get("profiles", {})
    schema_profile_required = set(
        catalog_schema.get("properties", {}).get("profiles", {}).get("required", [])
    )
    manifest_profile_required = set(
        manifest_schema.get("properties", {})
        .get("project", {}).get("properties", {})
        .get("profiles", {}).get("required", [])
    )
    if schema_profile_required != set(dimensions):
        errors.append("catalog schema profile dimensions disagree with catalog")
    if manifest_profile_required != set(dimensions):
        errors.append("manifest schema profile dimensions disagree with catalog")
    for dimension in dimensions:
        definitions = profiles.get(dimension, [])
        if not definitions:
            errors.append(f"{dimension}: profile registry must not be empty")
        orders = [item.get("order") for item in definitions]
        if len(set(orders)) != len(orders) or not all(isinstance(item, int) for item in orders):
            errors.append(f"{dimension}: profile order values must be unique integers")
        for item in definitions:
            identifier = item.get("id")
            for signal in item.get("signals", []):
                kind = signal.get("kind")
                if kind not in {"path", "content", "dependency"}:
                    errors.append(f"{dimension}/{identifier}: invalid signal kind")
                if kind in {"path", "content"} and (
                    not isinstance(signal.get("pattern"), str) or not signal.get("pattern")
                ):
                    errors.append(f"{dimension}/{identifier}: signal needs a pattern")
                if kind == "content" and not signal.get("contains"):
                    errors.append(f"{dimension}/{identifier}: content signal needs contains")
                if kind == "dependency" and (
                    not signal.get("ecosystem") or not signal.get("name")
                ):
                    errors.append(f"{dimension}/{identifier}: dependency signal needs ecosystem and name")
                strength = signal.get("strength")
                if strength is not None and strength not in {"strong", "weak"}:
                    errors.append(f"{dimension}/{identifier}: signal strength must be strong|weak")
                weight = signal.get("weight")
                if weight is not None and (
                    not isinstance(weight, (int, float)) or weight <= 0 or weight > 1
                ):
                    errors.append(f"{dimension}/{identifier}: signal weight must be in (0, 1]")
    cross_aliases: dict[str, list[str]] = {}
    for dimension in dimensions:
        for item in profiles.get(dimension, []):
            for name in [item.get("id"), *item.get("aliases", [])]:
                if isinstance(name, str):
                    cross_aliases.setdefault(name, []).append(f"{dimension}:{item.get('id')}")
    for name, owners in sorted(cross_aliases.items()):
        if len(owners) > 1:
            errors.append(f"cross-dimension profile name collision {name}: {', '.join(owners)}")
    for hint in catalog.get("cue_hints", []):
        if not isinstance(hint, dict) or not hint.get("cue") or not hint.get("note"):
            errors.append("cue_hints entries require cue and note")
    gate_schema_path = metadata / "discovery-gate-schema.json"
    if not gate_schema_path.is_file():
        errors.append("discovery-gate-schema.json is missing")
    else:
        gate_schema = read_json(gate_schema_path)
        if "judgment" not in gate_schema.get("definitions", {}):
            errors.append("discovery-gate-schema.json must define judgment")
        if "pack" not in gate_schema.get("definitions", {}):
            errors.append("discovery-gate-schema.json must define pack")
    for index, doc in enumerate(catalog.get("documents", [])):
        label = doc.get("id", f"document[{index}]")
        selection = doc.get("selection", {})
        obsolete = {"overlays", "include_if_overlay"} & set(selection)
        if obsolete:
            errors.append(f"{label}: obsolete selection fields: {', '.join(sorted(obsolete))}")
    # Drive the template scan from the catalog itself (template_file paths
    # now vary by group/kind under content/) plus the two shared templates
    # with no catalog reference (audit-report, audience-deepdive).
    template_paths = {
        SKILL_ROOT / detail["template_file"]
        for detail in query_catalog.load_all_types()
        if detail["template_file"].endswith(".md")
    }
    template_paths |= {
        SKILL_ROOT / "content" / "shared" / name
        for name in ("audit-report.template.md", "audience-deepdive.template.md")
    }
    for template in sorted(template_paths):
        if template.name in MARKDOWN_EXCEPTIONS:
            continue
        text = template.read_text(encoding="utf-8")
        if not text.startswith("---\ndocforge_provenance:\n"):
            errors.append(f"{template.name}: provenance frontmatter must be YAML docforge_provenance at byte one")
            continue
        state, provenance, _ = parse_frontmatter(text)
        if state != "ok":
            errors.append(f"{template.name}: provenance frontmatter state is {state}")
            continue
        if not isinstance(provenance, dict):
            errors.append(f"{template.name}: provenance frontmatter is not valid YAML")
            continue
        missing = sorted(PROVENANCE_FIELDS - set(provenance))
        graph = provenance.get("graph")
        generator = provenance.get("generator")
        if missing or not isinstance(graph, dict) or not {"provider", "flow"} <= set(graph):
            errors.append(f"{template.name}: provenance frontmatter is missing required fields")
        if not isinstance(generator, dict) or not {"name", "version"} <= set(generator):
            errors.append(f"{template.name}: provenance frontmatter is missing generator")
        if provenance.get("schema") != SCHEMA_VERSION or "graph_snapshot" in provenance:
            errors.append(f"{template.name}: provenance frontmatter must use schema 2.0")
    cli_py = SKILL_ROOT / "runtime" / "cli" / "python"
    cli_js = SKILL_ROOT / "runtime" / "cli" / "js"
    # One-shot migration helpers; not part of the dual-runtime public surface.
    peerless = {"split_catalog", "split_document_catalog"}
    py_names = {path.stem for path in cli_py.glob("*.py")} - peerless
    js_names = {path.stem for path in cli_js.glob("*.js")} - peerless
    for name in sorted(py_names - js_names):
        errors.append(f"missing Node peer for {name}.py")
    for name in sorted(js_names - py_names):
        errors.append(f"missing Python peer for {name}.js")
    runtime_root = SKILL_ROOT / "runtime"

    def impl_matches(name: str, suffix: str) -> list[Path]:
        return [
            path
            for path in runtime_root.rglob(f"{name}.{suffix}")
            if "cli" not in path.relative_to(runtime_root).parts
        ]

    for name, tokens in PUBLIC_CONTRACTS.items():
        for suffix in ("py", "js"):
            # CLI flags live in the runtime implementation; the cli/
            # launcher is a thin re-export with no flag text.
            matches = impl_matches(name, suffix)
            fallback = (cli_py if suffix == "py" else cli_js) / f"{name}.{suffix}"
            source = matches[0] if matches else fallback
            text = source.read_text(encoding="utf-8")
            missing = [token for token in tokens if token not in text]
            if missing:
                errors.append(f"{name}.{suffix}: missing CLI contract tokens: {', '.join(missing)}")
    plugin = read_json(REPO_ROOT / ".claude-plugin" / "plugin.json")
    market = read_json(REPO_ROOT / ".claude-plugin" / "marketplace.json")["plugins"][0]
    versions = {plugin.get("version"), market.get("version"), catalog.get("version")}
    if versions != {CATALOG_VERSION}:
        errors.append(f"release versions disagree: {sorted(str(item) for item in versions)}")
    skill_text = (REPO_ROOT / "skills" / "docforge" / "SKILL.md").read_text(encoding="utf-8")
    skill_match = re.search(r"^description: (.+)$", skill_text, re.MULTILINE)
    revise_text = (REPO_ROOT / "skills" / "docforge-revise" / "SKILL.md").read_text(encoding="utf-8")
    revise_match = re.search(r"^description: (.+)$", revise_text, re.MULTILINE)
    if not skill_match:
        errors.append("skills/docforge/SKILL.md missing description frontmatter")
    if not revise_match:
        errors.append("skills/docforge-revise/SKILL.md missing description frontmatter")
    descriptions = {
        plugin.get("description"),
        market.get("description"),
        skill_match.group(1) if skill_match else None,
    }
    if len(descriptions) != 1:
        errors.append("package descriptions disagree (plugin, marketplace, docforge SKILL.md)")
    if (REPO_ROOT / "meta.json").exists():
        errors.append("obsolete meta.json remains; Agent Skills install uses SKILL.md only")
    forbidden_files = {
        "document" + "-templates.json",
        "generation" + "-status.json",
        "status" + "-schema.json",
        "template" + "-schema.json",
    }
    present = forbidden_files & {path.name for path in metadata.iterdir()}
    if present:
        errors.append(f"obsolete metadata files remain: {', '.join(sorted(present))}")
    legacy_constants = ["SP" + "INE", "SPINE_" + "PLAN", "OVER" + "LAYS"]
    duplicate_constants = re.compile(r"\b(" + "|".join(legacy_constants) + r")\s*=")
    scannable = (
        list(cli_py.glob("*.py"))
        + list(cli_js.glob("*.js"))
        + [path for path in runtime_root.rglob("*.[pj][ys]") if "cli" not in path.relative_to(runtime_root).parts]
    )
    for script in sorted(scannable):
        if duplicate_constants.search(script.read_text(encoding="utf-8", errors="ignore")):
            errors.append(f"{script.name}: duplicated registry constant")
    return errors


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.parse_args()
    errors = validate()
    if errors:
        for error in errors:
            print(f"ERROR  {error}")
        print(f"\n{len(errors)} metadata errors.")
        return 1
    print("OK  catalog, schemas, templates, runtime peers, and package metadata agree.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/docforge/_shared/runtime/validation/validate_metadata.js
#!/usr/bin/env node
"use strict";
/** Validate the Docforge catalog, schemas, templates, peers, and release metadata. */

const fs = require("fs");
const path = require("path");
const { readJson } = require("../common/_util.js");
const pf = require("../common/provenance_frontmatter.js");
const { SPECIAL_DOC_SOURCES } = require("../common/special_files.js");
const queryCatalog = require("../catalog/query_catalog.js");
const generateIndexes = require("./generate_indexes.js");

const SKILL_ROOT = path.resolve(__dirname, "..", "..");
const REPO_ROOT = path.resolve(SKILL_ROOT, "..", "..", "..");
const EXCEPTIONS = SPECIAL_DOC_SOURCES;
const CATALOG_VERSION = "2.8.0";
const PUBLIC_CONTRACTS = {
  manage_manifest: ["init", "add", "set", "status", "audit", "reconcile", "--repo", "--tier", "--shape", "--platform", "--framework", "--concern", "--audience", "--type", "--id", "--path", "--evidence", "--status", "--mode", "--verdict", "--report"],
  detect_profiles: ["--repo", "--json", "--emit-gate-pack", "confirmed", "candidate"],
  scaffold_docs: ["--repo", "--manifest", "--dry-run", "--document", "--audit", "--revise"],
  precheck_graph: ["--repo", "--need", "code", "flow"],
  check_staleness: ["--manifest", "--document", "--section", "--json", "--sync-provenance"],
  flow_index: ["harvest", "revise", "render", "organize", "emit", "apply", "--repo", "--gitnexus-export", "--main-limit", "--output", "--organization"],
  migrate_metadata: ["--repo", "--manifest", "--dry-run", "--report"],
  query_catalog: ["--tier", "--id", "--ids", "--profile", "--applicable", "--validate", "--category", "--route"],
  generate_indexes: ["--write", "--check"],
};

function validate() {
  const errors = [];
  const metadata = path.join(SKILL_ROOT, ".metadata");
  errors.push(...queryCatalog.validate());
  try {
    for (const target of generateIndexes.staleTargets()) {
      errors.push(`generated router is stale: ${path.relative(SKILL_ROOT, target)} (run generate_indexes --write)`);
    }
  } catch (error) {
    errors.push(`generate_indexes --check failed: ${error.message}`);
  }
  let catalog;
  try {
    catalog = queryCatalog.asLegacyCatalog();
  } catch (error) {
    errors.push(error.message);
    catalog = { version: null, documents: [], profiles: {}, tiers: [] };
  }
  const catalogSchema = readJson(path.join(metadata, "catalog-schema.json"));
  const manifestSchema = readJson(path.join(metadata, "manifest-schema.json"));
  const flowIndexSchema = readJson(path.join(metadata, "flow-index-schema.json"));
  const provenanceSchemaPath = path.join(metadata, "provenance-schema.json");
  if (!fs.existsSync(provenanceSchemaPath)) {
    errors.push("provenance-schema.json is missing");
  } else {
    const provenanceSchema = readJson(provenanceSchemaPath);
    if ((((provenanceSchema.properties || {}).schema || {}).const) !== pf.SCHEMA_VERSION) {
      errors.push("provenance schema must require schema 2.0");
    }
  }
  if (catalog.version !== CATALOG_VERSION) errors.push(`catalog version must be ${CATALOG_VERSION}`);
  if ((((catalogSchema.properties || {}).version || {}).const) !== CATALOG_VERSION) {
    errors.push("catalog schema version disagrees with catalog");
  }
  if (!fs.existsSync(path.join(metadata, "catalog", "index.json"))) {
    errors.push("split catalog index.json is missing");
  }
  if (fs.existsSync(path.join(metadata, "catalog.json"))) {
    errors.push("obsolete monolith catalog.json remains; use .metadata/catalog/");
  }
  if ((((manifestSchema.properties || {}).version || {}).const) !== "3.1") {
    errors.push("manifest schema must require version 3.1");
  }
  if ((((flowIndexSchema.properties || {}).version || {}).const) !== "1.1") {
    errors.push("flow index schema must require version 1.1");
  }
  const flowItem = ((((flowIndexSchema.properties || {}).flows || {}).items || {}).properties) || {};
  for (const field of ["display_name", "family", "doc_role", "composed_into", "doc_path"]) {
    if (!(field in flowItem)) errors.push(`flow index schema must define flow.${field}`);
  }
  const docRoles = ((flowItem.doc_role || {}).enum) || [];
  if (docRoles.length !== 3 || !["standalone", "member", "index_only"].every((item) => docRoles.includes(item))) {
    errors.push("flow index schema doc_role must be standalone|member|index_only");
  }
  const dimensions = ["shapes", "platforms", "frameworks", "concerns", "audiences"];
  const schemaProfileRequired = new Set(((((catalogSchema.properties || {}).profiles || {}).required) || []));
  const manifestProfileRequired = new Set(((((((manifestSchema.properties || {}).project || {}).properties || {}).profiles || {}).required) || []));
  if (schemaProfileRequired.size !== dimensions.length || dimensions.some((item) => !schemaProfileRequired.has(item))) {
    errors.push("catalog schema profile dimensions disagree with catalog");
  }
  if (manifestProfileRequired.size !== dimensions.length || dimensions.some((item) => !manifestProfileRequired.has(item))) {
    errors.push("manifest schema profile dimensions disagree with catalog");
  }
  for (const dimension of dimensions) {
    const definitions = (catalog.profiles || {})[dimension] || [];
    if (!definitions.length) errors.push(`${dimension}: profile registry must not be empty`);
    const orders = definitions.map((item) => item.order);
    if (new Set(orders).size !== orders.length || orders.some((item) => !Number.isInteger(item))) {
      errors.push(`${dimension}: profile order values must be unique integers`);
    }
    for (const item of definitions) {
      for (const signal of item.signals || []) {
        if (!["path", "content", "dependency"].includes(signal.kind)) {
          errors.push(`${dimension}/${item.id}: invalid signal kind`);
        }
        if (["path", "content"].includes(signal.kind) && (typeof signal.pattern !== "string" || !signal.pattern)) {
          errors.push(`${dimension}/${item.id}: signal needs a pattern`);
        }
        if (signal.kind === "content" && !signal.contains) {
          errors.push(`${dimension}/${item.id}: content signal needs contains`);
        }
        if (signal.kind === "dependency" && (!signal.ecosystem || !signal.name)) {
          errors.push(`${dimension}/${item.id}: dependency signal needs ecosystem and name`);
        }
        if (signal.strength != null && !["strong", "weak"].includes(signal.strength)) {
          errors.push(`${dimension}/${item.id}: signal strength must be strong|weak`);
        }
        if (signal.weight != null && (typeof signal.weight !== "number" || signal.weight <= 0 || signal.weight > 1)) {
          errors.push(`${dimension}/${item.id}: signal weight must be in (0, 1]`);
        }
      }
    }
  }
  const crossAliases = new Map();
  for (const dimension of dimensions) {
    for (const item of (catalog.profiles || {})[dimension] || []) {
      for (const name of [item.id, ...(item.aliases || [])]) {
        if (typeof name !== "string") continue;
        if (!crossAliases.has(name)) crossAliases.set(name, []);
        crossAliases.get(name).push(`${dimension}:${item.id}`);
      }
    }
  }
  for (const name of [...crossAliases.keys()].sort()) {
    const owners = crossAliases.get(name);
    if (owners.length > 1) errors.push(`cross-dimension profile name collision ${name}: ${owners.join(", ")}`);
  }
  for (const hint of catalog.cue_hints || []) {
    if (!hint || typeof hint !== "object" || !hint.cue || !hint.note) {
      errors.push("cue_hints entries require cue and note");
    }
  }
  const gateSchemaPath = path.join(metadata, "discovery-gate-schema.json");
  if (!fs.existsSync(gateSchemaPath)) errors.push("discovery-gate-schema.json is missing");
  else {
    const gateSchema = readJson(gateSchemaPath);
    if (!(((gateSchema.definitions || {}).judgment))) errors.push("discovery-gate-schema.json must define judgment");
    if (!(((gateSchema.definitions || {}).pack))) errors.push("discovery-gate-schema.json must define pack");
  }
  for (let index = 0; index < (catalog.documents || []).length; index++) {
    const doc = catalog.documents[index];
    const label = doc.id || `document[${index}]`;
    const selection = doc.selection || {};
    const obsolete = ["overlays", "include_if_overlay"].filter((field) => field in selection).sort();
    if (obsolete.length) errors.push(`${label}: obsolete selection fields: ${obsolete.join(", ")}`);
  }
  // Drive the template scan from the catalog itself (template_file paths
  // now vary by group/kind under content/) plus the two shared templates
  // with no catalog reference (audit-report, audience-deepdive).
  const templatePaths = new Set(
    queryCatalog
      .loadAllTypes()
      .map((detail) => detail.template_file)
      .filter((rel) => rel.endsWith(".md"))
      .map((rel) => path.join(SKILL_ROOT, rel)),
  );
  for (const name of ["audit-report.template.md", "audience-deepdive.template.md"]) {
    templatePaths.add(path.join(SKILL_ROOT, "content", "shared", name));
  }
  for (const templatePath of [...templatePaths].sort()) {
    const name = path.basename(templatePath);
    if (EXCEPTIONS.has(name)) continue;
    const text = fs.readFileSync(templatePath, "utf8");
    if (!text.startsWith("---\ndocforge_provenance:\n")) {
      errors.push(`${name}: provenance frontmatter must be YAML docforge_provenance at byte one`);
      continue;
    }
    const parsed = pf.parseFrontmatter(text);
    if (parsed.state !== "ok") {
      errors.push(`${name}: provenance frontmatter state is ${parsed.state}`);
      continue;
    }
    const provenance = parsed.provenance;
    if (!provenance || typeof provenance !== "object" || Array.isArray(provenance)) {
      errors.push(`${name}: provenance frontmatter is not valid YAML`);
      continue;
    }
    const missing = [...pf.PROVENANCE_FIELDS].filter((key) => !(key in provenance));
    const graph = provenance.graph;
    const generator = provenance.generator;
    if (missing.length || !graph || typeof graph !== "object" || !("provider" in graph) || !("flow" in graph)) {
      errors.push(`${name}: provenance frontmatter is missing required fields`);
    }
    if (!generator || typeof generator !== "object" || !("name" in generator) || !("version" in generator)) {
      errors.push(`${name}: provenance frontmatter is missing generator`);
    }
    if (provenance.schema !== pf.SCHEMA_VERSION || "graph_snapshot" in provenance) {
      errors.push(`${name}: provenance frontmatter must use schema 2.0`);
    }
  }
  const cliPy = path.join(SKILL_ROOT, "runtime", "cli", "python");
  const cliJs = path.join(SKILL_ROOT, "runtime", "cli", "js");
  const peerless = new Set(["split_catalog", "split_document_catalog"]);
  const py = new Set(
    fs.readdirSync(cliPy)
      .filter((name) => name.endsWith(".py"))
      .map((name) => path.basename(name, ".py"))
      .filter((name) => !peerless.has(name)),
  );
  const js = new Set(
    fs.readdirSync(cliJs)
      .filter((name) => name.endsWith(".js"))
      .map((name) => path.basename(name, ".js"))
      .filter((name) => !peerless.has(name)),
  );
  for (const name of [...py].filter((name) => !js.has(name)).sort()) errors.push(`missing Node peer for ${name}.py`);
  for (const name of [...js].filter((name) => !py.has(name)).sort()) errors.push(`missing Python peer for ${name}.js`);
  const runtimeRoot = path.join(SKILL_ROOT, "runtime");
  function findImpl(dir, filename, relativeParts = []) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (entry.name === "cli" && relativeParts.length === 0) continue;
        const found = findImpl(full, filename, relativeParts.concat(entry.name));
        if (found) return found;
      } else if (entry.name === filename) {
        return full;
      }
    }
    return null;
  }
  for (const [name, tokens] of Object.entries(PUBLIC_CONTRACTS)) {
    for (const suffix of ["py", "js"]) {
      // CLI flags live in the runtime implementation; the cli/ launcher
      // is a thin re-export with no flag text.
      const fallback = path.join(suffix === "py" ? cliPy : cliJs, `${name}.${suffix}`);
      const source = findImpl(runtimeRoot, `${name}.${suffix}`) || fallback;
      const text = fs.readFileSync(source, "utf8");
      const missing = tokens.filter((token) => !text.includes(token));
      if (missing.length) errors.push(`${name}.${suffix}: missing CLI contract tokens: ${missing.join(", ")}`);
    }
  }
  const plugin = readJson(path.join(REPO_ROOT, ".claude-plugin", "plugin.json"));
  const market = readJson(path.join(REPO_ROOT, ".claude-plugin", "marketplace.json")).plugins[0];
  const versions = new Set([plugin.version, market.version, catalog.version]);
  if (versions.size !== 1 || !versions.has(CATALOG_VERSION)) {
    errors.push(`release versions disagree: ${[...versions].map(String).sort().join(", ")}`);
  }
  const skillText = fs.readFileSync(path.join(REPO_ROOT, "skills", "docforge", "SKILL.md"), "utf8");
  const skillMatch = skillText.match(/^description: (.+)$/m);
  const reviseText = fs.readFileSync(path.join(REPO_ROOT, "skills", "docforge-revise", "SKILL.md"), "utf8");
  const reviseMatch = reviseText.match(/^description: (.+)$/m);
  if (!skillMatch) errors.push("skills/docforge/SKILL.md missing description frontmatter");
  if (!reviseMatch) errors.push("skills/docforge-revise/SKILL.md missing description frontmatter");
  if (new Set([plugin.description, market.description, skillMatch ? skillMatch[1] : null]).size !== 1) {
    errors.push("package descriptions disagree (plugin, marketplace, docforge SKILL.md)");
  }
  if (fs.existsSync(path.join(REPO_ROOT, "meta.json"))) {
    errors.push("obsolete meta.json remains; Agent Skills install uses SKILL.md only");
  }
  const IGNORED_DIRS = new Set([".git", ".pytest_cache", "__pycache__", ".venv", "venv", "node_modules"]);
  function collectReadmes(dir, out) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory() && !IGNORED_DIRS.has(entry.name)) collectReadmes(full, out);
      else if (entry.isFile() && entry.name === "README.md" && full !== path.join(REPO_ROOT, "README.md")) {
        out.push(path.relative(REPO_ROOT, full).split(path.sep).join("/"));
      }
    }
  }
  const readmes = []; collectReadmes(REPO_ROOT, readmes);
  if (readmes.length) errors.push(`nested README.md files are obsolete; use README.md: ${readmes.sort().join(", ")}`);
  const forbidden = new Set(["document" + "-templates.json", "generation" + "-status.json", "status" + "-schema.json", "template" + "-schema.json"]);
  const present = fs.readdirSync(metadata).filter((name) => forbidden.has(name)).sort();
  if (present.length) errors.push(`obsolete metadata files remain: ${present.join(", ")}`);
  const legacyConstants = ["SP" + "INE", "SPINE_" + "PLAN", "OVER" + "LAYS"];
  const duplicate = new RegExp(`\\b(${legacyConstants.join("|")})\\s*=`);
  function collectSourceFiles(dir, out, relativeParts = []) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (entry.name === "cli" && relativeParts.length === 0) {
          collectSourceFiles(path.join(full, "python"), out, relativeParts.concat("cli", "python"));
          collectSourceFiles(path.join(full, "js"), out, relativeParts.concat("cli", "js"));
          continue;
        }
        collectSourceFiles(full, out, relativeParts.concat(entry.name));
      } else if (/\.(py|js)$/.test(entry.name)) {
        out.push(full);
      }
    }
  }
  const scannable = [];
  collectSourceFiles(runtimeRoot, scannable);
  for (const source of scannable.sort()) {
    if (duplicate.test(fs.readFileSync(source, "utf8"))) {
      errors.push(`${path.basename(source)}: duplicated registry constant`);
    }
  }
  return errors;
}

function main() {
  const args = process.argv.slice(2);
  if (args.some((arg) => !["-h", "--help"].includes(arg))) {
    console.error(`error: unknown option: ${args.find((arg) => !["-h", "--help"].includes(arg))}`);
    return 2;
  }
  if (args.length) {
    console.log("usage: validate_metadata.js");
    return 0;
  }
  const errors = validate();
  if (errors.length) {
    for (const error of errors) console.log(`ERROR  ${error}`);
    console.log(`\n${errors.length} metadata errors.`);
    return 1;
  }
  console.log("OK  catalog, schemas, templates, runtime peers, and package metadata agree.");
  return 0;
}

process.exit(main());
